stablehlo/dialect/TypeInference.cpp

```diff
@@ -2390,11 +2390,26 @@ LogicalResult inferSliceOp(Optional<Location> location, Value operand,
   return success();
 }
 
-LogicalResult inferSortOp(
-    Optional<Location>, ValueRange inputs,
-    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
-  for (auto resultType : inputs.getTypes())
-    inferredReturnShapes.emplace_back(resultType.cast<ShapedType>());
+LogicalResult inferSortOp(Optional<Location> location, ValueRange inputs,
+                          SmallVectorImpl<Type>& inferredReturnTypes) {
+  // Infermost specific type from all operands
+  SmallVector<Type> mostSpecificType;
+  if (failed(hlo::inferMostSpecificType(location, inputs.getTypes(),
+                                        mostSpecificType))) {
+    return failure();
+  }
+
+  auto rankedOutput = mostSpecificType.front().dyn_cast<RankedTensorType>();
+  if (!rankedOutput) {
+    for (auto input : inputs.getTypes())
+      inferredReturnTypes.emplace_back(input);
+    return success();
+  }
+
+  for (auto input : inputs.getTypes())
+    inferredReturnTypes.emplace_back(RankedTensorType::get(
+        rankedOutput.getShape(), input.dyn_cast<TensorType>().getElementType(),
+        rankedOutput.getEncoding()));
```

Comment:
This is a non-trivial change, and we'll need to document why we're making it. PR description might be a good place for that. On an unrelated note, the description currently links to a wrong ticket.